File: cpp/src/0160.cpp
/*
[160. Intersection of Two Linked Lists (Easy)](https://leetcode.com/problems/intersection-of-two-linked-lists/)

@topic: Hash Table
@topic: Linked List
@topic: Two Pointers

Constraints:
  - The number of nodes of `listA` is in the `m`.
  - The number of nodes of `listB` is in the `n`.
  - `1 <= m, n <= 3 * 10^4`
  - `1 <= Node.val <= 10^5`
  - `0 <= skipA < m`
  - `0 <= skipB < n`
  - `intersectVal` is `0` if `listA` and `listB` do not intersect.
  - `intersectVal == listA[skipA] == listB[skipB]` if `listA` and `listB` intersect.
*/

#include "structs/LinkedList.hpp"

class Solution {
   public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if (!headA || !headB) return nullptr;
        auto p = headA;
        auto q = headB;
        while (p != q) {
            p = p ? p->next : headB;
            q = q ? q->next : headA;
        }
        return p;
    }
};

File: csharp/src/0160.cs
using System;
using System.Diagnostics;

using Structs;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode GetIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null) {
            return null;
        }
        var p = headA;
        var q = headB;
        while (p != q) {
            p = p?.next ?? headB;
            q = q?.next ?? headA;
        }
        return headA;
    }
}
